test: Tests for Metrics API enhancement to include error counters (#7423)
triton-inference-server/server: docs/user_guide/metrics.md

|              |Execution Count |`nv_inference_exec_count` |Number of inference batch executions (see [Inference Request Metrics](#inference-request-metrics), does not include cached requests)|Per model|Per request|
|              |Pending Request Count |`nv_inference_pending_request_count` |Number of inference requests awaiting execution by a backend. This number is incremented when a request is enqueued to the server (`TRITONSERVER_ServerInferAsync`) and is decremented when a backend is about to start executing the request. More details can be found below. |Per model|Per request|

> #### Failure Count is further classified into four distinct types:

Reviewer: ```suggestion
#### Failure Count Categories

The `nv_inference_request_failure` metric is further categorized to track the types of failures encountered through the use of a `reason` label, described below.

> |Sub-Category      |Sub-Metric | Sub-Metric Name |Description |

Reviewer: I think this can be condensed to 2 columns for now, something like:
---

|Failed Request Reason | Description |
|---|---|
| REJECTED | Number of ... |
| CANCELLED | Number of ... |
| BACKEND | Number of ... |
| OTHER | Number of ... |

|--------------|----------------|------------|------------|
|Failure Reason         |Rejected |REJECTED  | Number of inference failures due to request timeout in the schedular. |
|              |Canceled |CANCELED  |  Number of inference failures due to request cancellation in the core. |

> |              |Backend  |BACKEND |  Number of inference failures due to errors from backend. |

Reviewer: ```suggestion
|              |Backend  |BACKEND |  Number of inference failures during execution of requests in the backend/model. |
```

> |              |Other | OTHER  | Number of inference failures due to unspecified reasons in the core. |

Reviewer: ```suggestion
|              |Other | OTHER  | Number of inference failures due to other uncategorized reasons in the core. |
```

> **Note**
>

> > Ensemble failure metrics will reflect the failure counts of their composing models as well as the parent model, but currently do not capture the same granularity for the "reason" label and will default to the "OTHER" reason.

Reviewer: Adding a few more details to try to get ahead of future customer questions/confusion.
```suggestion
> Ensemble failure metrics will reflect the failure counts of their composing models as well as the parent model, but currently do not capture the same granularity for the "reason" label and will default to the "OTHER" reason.
>
> For example, if EnsembleA contains ModelA, and ModelA experiences a failed request due to a queue/backlog timeout in the scheduler, ModelA will have a failed request metric reflecting `reason=REJECTED` and `count=1`.
> Additionally, EnsembleA will have a failed request metric reflecting `reason=OTHER` and `count=2`. 
> The `count=2` reflects 1 from the internally failed request captured by ModelA, as well as 1 from the failed top-level request sent to EnsembleA by the user/client.
> The `reason=OTHER` reflects that fact that the ensemble doesn't currently capture the specific reason why 
> ModelA's request failed at this time.
```
